readthedocs/proxito/views.py

```diff
@@ -0,0 +1,256 @@
+"""Views for doc serving."""
+
+import logging
+import mimetypes
+import os
+from functools import wraps
+from urllib.parse import urlparse
+
+from django.conf import settings
+from django.core.files.storage import get_storage_class
+from django.http import Http404, HttpResponse, HttpResponseRedirect
+from django.shortcuts import get_object_or_404, render
+from django.utils.encoding import iri_to_uri
+from django.views.static import serve
+
+from readthedocs.core.resolver import resolve
+from readthedocs.projects.models import Project, ProjectRelationship
+
+
+log = logging.getLogger(__name__)  # noqa
+
+
+def _serve_401(request, project):
+    res = render(request, '401.html')
+    res.status_code = 401
+    log.debug('Unauthorized access to %s documentation', project.slug)
+    return res
+
+
+def _fallback():
+    # TODO: This currently isn't used. It might be though, so keeping it for now.
+    res = HttpResponse('Internal fallback to RTD app')
+    res.status_code = 420
+    log.debug('Falling back to RTD app')
+    return res
+
+
+def map_subproject_slug(view_func):
+    """
+    A decorator that maps a ``subproject_slug`` URL param into a Project.
+
+    :raises: Http404 if the Project doesn't exist
+
+    .. warning:: Does not take into account any kind of privacy settings.
+    """
+
+    @wraps(view_func)
+    def inner_view(  # noqa
+            request, subproject=None, subproject_slug=None, *args, **kwargs
+    ):
+        if subproject is None and subproject_slug:
+            # Try to fetch by subproject alias first, otherwise we might end up
+            # redirected to an unrelated project.
+            # Depends on a project passed into kwargs
+            rel = ProjectRelationship.objects.filter(
+                parent=kwargs['project'],
+                alias=subproject_slug,
+            ).first()
+            if rel:
+                subproject = rel.child
+            else:
+                rel = ProjectRelationship.objects.filter(
+                    parent=kwargs['project'],
+                    child__slug=subproject_slug,
+                ).first()
+                if rel:
+                    subproject = rel.child
+                else:
+                    log.error('Invalid subproject slug. subproject_slug=%s', subproject_slug)
```

Comment:
Maybe this is not an error, but `warning`. Modifying any `/projects` URL will produce this error otherwise.

Also, I'd be clearer in the message:

```suggestion
                    log.warning('The slug is not subproject of project. subproject_slug=%s project_slug=%s', subproject_slug, kwargs['project'].slug)